File: nightgrass/src/main/java/com/goodfeel/nightgrass/service/StripeService.java
package com.goodfeel.nightgrass.service;

import com.stripe.Stripe;
import com.stripe.model.checkout.Session;
import com.stripe.param.checkout.SessionCreateParams;
import com.stripe.param.checkout.SessionCreateParams.PaymentMethodType;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

import java.math.BigDecimal;

@Service
public class StripeService {

    public StripeService(@Value("${stripe.secret-key}") String secretKey) {
        Stripe.apiKey = secretKey;
    }

    public Mono<Session> createCheckoutSession(BigDecimal amount, String currency, String successUrl, String cancelUrl) {
        return Mono.fromCallable(() -> {
            SessionCreateParams params = SessionCreateParams.builder()
                    .addPaymentMethodType(PaymentMethodType.CARD)
                    .setMode(SessionCreateParams.Mode.PAYMENT)
                    .setSuccessUrl(successUrl)
                    .setCancelUrl(cancelUrl)
                    .addLineItem(
                            SessionCreateParams.LineItem.builder()
                                    .setQuantity(1L)
                                    .setPriceData(
                                            SessionCreateParams.LineItem.PriceData.builder()
                                                    .setCurrency(currency)
                                                    .setUnitAmount(amount.multiply(BigDecimal.valueOf(100)).longValueExact()) // amount in cents
                                                    .setProductData(
                                                            SessionCreateParams.LineItem.PriceData.ProductData.builder()
                                                                    .setName("E-commerce Product")
                                                                    .build()
                                                    )
                                                    .build()
                                    )
                                    .build()
                    )
                    .build();

            return Session.create(params);
        });
    }
}

File: nightgrass/src/main/kotlin/com/goodfeel/nightgrass/service/StripeService.kt
package com.goodfeel.nightgrass.service

import com.stripe.Stripe
import com.stripe.model.checkout.Session
import com.stripe.param.checkout.SessionCreateParams
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono
import java.math.BigDecimal

@Service
class StripeService(@Value("\${stripe.secret-key}") secretKey: String) {
    init {
        Stripe.apiKey = secretKey
    }

    fun createCheckoutSession(
        amount: BigDecimal,
        currency: String,
        successUrl: String,
        cancelUrl: String
    ): Mono<Session> {
        return Mono.fromCallable {
            val params =
                SessionCreateParams.builder()
                    .addPaymentMethodType(SessionCreateParams.PaymentMethodType.CARD)
                    .setMode(SessionCreateParams.Mode.PAYMENT)
                    .setSuccessUrl(successUrl)
                    .setCancelUrl(cancelUrl)
                    .addLineItem(
                        SessionCreateParams.LineItem.builder()
                            .setQuantity(1L)
                            .setPriceData(
                                SessionCreateParams.LineItem.PriceData.builder()
                                    .setCurrency(currency)
                                    .setUnitAmount(
                                        amount.multiply(BigDecimal.valueOf(100)).longValueExact()
                                    ) // amount in cents
                                    .setProductData(
                                        SessionCreateParams.LineItem.PriceData.ProductData.builder()
                                            .setName("E-commerce Product")
                                            .build()
                                    )
                                    .build()
                            )
                            .build()
                    )
                    .build()
            Session.create(params)
        }
    }
}
